middleware/service_broker_rate_limiter.rb

```diff
@@ -0,0 +1,101 @@
+module CloudFoundry
+  module Middleware
+    class ServiceBrokerRequestCounter
+      include Singleton
+
+      def initialize
+        @mutex = Mutex.new
+        @data = {}
+      end
+
+      def can_make_request?(user_guid, limit)
+        @mutex.synchronize do
+          request_count = @data.fetch(user_guid, 0)
+          return false if request_count + 1 > limit
+
+          @data[user_guid] = request_count + 1
+          true
+        end
+      end
+
+      def decrement(user_guid)
+        @mutex.synchronize do
+          @data[user_guid] = @data[user_guid] - 1
+        end
+      end
+    end
+
+    class ServiceBrokerRateLimiter
+      def initialize(app, logger:, concurrent_limit:)
+        @app                               = app
+        @logger                            = logger
+        @concurrent_limit = concurrent_limit
+        @request_counter = ServiceBrokerRequestCounter.instance
+      end
+
+      def call(env)
+        request = ActionDispatch::Request.new(env)
+        user_guid = env['cf.user_guid']
+
+        unless skip_rate_limiting?(env, request)
+          return too_many_requests!(env) unless @request_counter.can_make_request?(user_guid, @concurrent_limit)
+
+          begin
+            return @app.call(env)
+          rescue => e
+            raise e
+          ensure
+            @request_counter.decrement(user_guid)
+          end
+        end
+
+        @app.call(env)
+      end
+
+      private
+
+      def skip_rate_limiting?(env, request)
+        return admin? || !is_service_request?(request) || !rate_limit_method?(request)
+      end
+
+      def admin?
+        VCAP::CloudController::SecurityContext.admin? || VCAP::CloudController::SecurityContext.admin_read_only?
+      end
+
+      def is_service_request?(request)
+        [
+          %r{\A/v2/service_instances},
+          %r{\A/v2/service_bindings},
+          %r{\A/v2/service_keys},
+          %r{\A/v3/service_instances},
```

Comment:
Lets say we have this call POST /v3/service_instances/:guid/relationships/shared_spaces , i would assume it will count the counter yet afaik the endpoint doesnt make a broker request. Do we just say its sufficient to wildcard match the whole  /v3/service_instances endpoint or do we want a more finegrained list so specify each endpoint explicitely ?

See https://v3-apidocs.cloudfoundry.org/version/3.110.0/index.html#share-a-service-instance-to-other-spaces